modules/core/include/opencv2/core/hal/cvtype.h

```diff
@@ -0,0 +1,256 @@
+#define CV_MAT_CN_MASK            (CV_SANITY_CN_MASK    << CV_CN_SHIFT   )
+#define CV_MAT_TYPE_MASK          (CV_MAT_CN_MASK | CV_MAT_DEPTH_MASK)
+
+//ElemType register limits (public)
+#define CV_DEPTH_MAX              (1 << CV_DEPTH_LEN)
+#define CV_CN_MAX                 (1 << (CV_CN_BASE_LEN + CV_SANITY_CN_EXP_MASK)) /*(1 << CV_CN_LEN )*/
+
+//ElemType register deflate/inflate operations (private)
+#define __CV_CN_MAKE_EXPONENT(cn) static_cast<int>((~static_cast<int>((cn) > CV_SANITY_CN_BASE_MASK) + 1) & \
+                                      (0x45263107U >> (((((cn) & -static_cast<int>(cn)) * 0x17) >> 2) & 0x1C)) & CV_SANITY_CN_EXP_MASK)
+#define __CV_CN_EXPONENT(cn_bin)  (((cn_bin) >> CV_CN_BASE_LEN) & CV_SANITY_CN_EXP_MASK)
+#define __CV_CN_MAKE_BASE(cn)     (((cn) >> __CV_CN_MAKE_EXPONENT(cn))-1) /* do not do CV_SANITY_CN_BASE_MASK here */
+#define __CV_CN_BASE(cn_bin)      ((cn_bin) & CV_SANITY_CN_BASE_MASK)
+#define __CV_CN_DEFLATE(cn)       ((__CV_CN_MAKE_EXPONENT(cn) << CV_CN_BASE_LEN) | (__CV_CN_MAKE_BASE(cn) & CV_SANITY_CN_BASE_MASK)) /*((cn) - 1)*/
+#define __CV_CN_INFLATE(cn_bin)   ((__CV_CN_BASE(cn_bin) + 1) << __CV_CN_EXPONENT(cn_bin)) /*((cn_bin) + 1)*/
+
+//ElemType register pack/unpack operations (private)
+#define __CV_DEPTH_PACK(depth)    ((depth & CV_SANITY_DEPTH_MASK) << CV_DEPTH_SHIFT)
+#define __CV_DEPTH_UNPACK(flags)  (((flags) & CV_MAT_DEPTH_MASK)  >> CV_DEPTH_SHIFT)
+#define __CV_CN_PACK(cn)          ((__CV_CN_DEFLATE(cn) & CV_SANITY_CN_MASK) << CV_CN_SHIFT)
+#define __CV_CN_UNPACK(flags)     __CV_CN_INFLATE(((flags) & CV_MAT_CN_MASK) >> CV_CN_SHIFT)
+/*
+TODO: static assertion on bad input channel
+TODO: static assertion on cn == cn to avoid random multiple evaluation*/
+#define __CV_TYPE_PACK(depth,cn)  (__CV_DEPTH_PACK(depth) | __CV_CN_PACK(cn))
+/*TODO: Rewite __CV_TYPE_UNPACK*/
+#define __CV_TYPE_UNPACK(flags)   __CV_TYPE_PACK(__CV_DEPTH_UNPACK(flags),__CV_CN_UNPACK(flags)) /*((flags) & CV_MAT_TYPE_MASK)*/
+
+//ElemType register - user operations (public)
+#define CV_CN_ISVALID(cn)         (cn > 0 && cn <= CV_CN_MAX && __CV_CN_MAKE_BASE(cn) <= CV_SANITY_CN_BASE_MASK)
+#define CV_MAT_TYPE(flags)        __CV_TYPE_UNPACK(flags)
+#define CV_MAT_NO_TYPE(flags)     ((flags) & ~CV_MAT_TYPE_MASK)
+#define CV_MAT_CN(flags)          __CV_CN_UNPACK(flags)
+#define CV_MAT_DEPTH(flags)       __CV_DEPTH_UNPACK(flags)
+#define CV_MAKETYPE(depth,cn)     __CV_TYPE_PACK(depth,cn)
+#define CV_MAKE_TYPE              CV_MAKETYPE
+
+#define CV_8U        0
+#define CV_16U       2
+#define CV_32U       8
+#define CV_64U       9
+
+#define CV_8S        1
+#define CV_16S       3
+#define CV_32S       4
+#define CV_64S       11
+
+#define CV_16F       7
+#define CV_32F       5  /*avoid using this - order broken!*/
+#define CV_64F       6  /*avoid using this - order broken!*/
+#define CV_32F_      13 /*restore order among IEEE floats*/
+#define CV_64F_      14 /*restore order among IEEE floats*/
+
+#define CV_8Q        17
```

Comment:
what's that? fixed-point? there is no commonly used fixed-point types; the number of fractional bits varies a lot from one algorithm to another. Let's remove it.

Reply:
Please refer to:
- https://github.com/opencv/opencv/pull/12611/files#diff-39872d2057ab847614b5bb242a81207bR339
- https://github.com/opencv/opencv/pull/12611/files#diff-93b27ad7ebe8d837ba1a6bc86636e131R789

Reply:
Kindly let us discuss that in #12643

Reply:
Any decision about Q formats?